File: example1/src/main/java/com/example/grpc/cancellation/CancellationServerStreamingServer.java
package com.example.grpc.cancellation;

import com.example.grpc.Delays;
import com.example.grpc.EchoRequest;
import com.example.grpc.EchoResponse;
import com.example.grpc.EchoServiceGrpc;
import com.example.grpc.Servers;
import io.grpc.Status;
import io.grpc.stub.ServerCallStreamObserver;
import io.grpc.stub.StreamObserver;

import java.util.logging.Level;
import java.util.logging.Logger;

public class CancellationServerStreamingServer {

    private static final Logger logger = Logger.getLogger(CancellationServerStreamingServer.class.getName());

    public static void main(String[] args) throws Exception {
        Servers.start(new EchoServiceImpl());
    }

    private static class EchoServiceImpl extends EchoServiceGrpc.EchoServiceImplBase {
        @Override
        public void serverStreamingEcho(EchoRequest request, StreamObserver<EchoResponse> responseObserver) {
            logger.log(Level.INFO, "request: {0}", request.getMessage());
            var serverObserver = (ServerCallStreamObserver<EchoResponse>) responseObserver;
            for (var i = 0; i <= 9; i++) {
                if (serverObserver.isCancelled()) {
                    logger.info("server received cancellation");
                    responseObserver.onError(Status.CANCELLED.withDescription("Server confirmed cancellation").asRuntimeException());
                    return;
                }

                Delays.sleep(1);

                var response = EchoResponse.newBuilder().setMessage("hello " + request.getMessage() + " " + i).build();
                responseObserver.onNext(response);
            }
            responseObserver.onCompleted();
        }
    }
}

File: grpc-kotlin-examples/src/main/kotlin/com/example/grpc/cancellation/CancellationServerStreamingServer.kt
package com.example.grpc.cancellation

import com.example.grpc.EchoRequest
import com.example.grpc.EchoResponse
import com.example.grpc.EchoServiceGrpcKt
import com.example.grpc.Servers
import com.example.grpc.echoResponse
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.*
import java.util.logging.Logger

object CancellationServerStreamingServer {
  private val logger = Logger.getLogger(CancellationServerStreamingServer::class.java.name)

  @JvmStatic
  fun main(args: Array<String>) {
    Servers.start(EchoServiceImpl())
  }

  private class EchoServiceImpl : EchoServiceGrpcKt.EchoServiceCoroutineImplBase() {
    override fun serverStreamingEcho(request: EchoRequest): Flow<EchoResponse> {
      val name = request.message
      logger.info("request: $name")

      try {
        for (i in 0..9) {
          // Cooperative cancellation: emit() and delay() check for cancellation automatically
          emit(echoResponse { message = "hello $name $i" })

          Delays.sleep(1)
        }
      } catch (e: CancellationException) {
        logger.info("server received cancellation")
        // Re-throwing is standard to allow gRPC to finalize the RPC status
        throw e
      }
    }
  }
}
